curator-recipes/src/main/java/org/apache/curator/framework/recipes/leader/LeaderSelector.java

```diff
@@ -237,13 +242,15 @@ private synchronized boolean internalRequeue()
         if ( !isQueued && (state.get() == State.STARTED) )
         {
             isQueued = true;
-            Future<Void> task = executorService.submit(new Callable<Void>()
+            Future<?> oldTask = ourTask;
+            ourTask = executorService.submit(new Callable<Void>()
             {
                 @Override
                 public Void call() throws Exception
                 {
                     try
                     {
+                        waitTaskDone(oldTask, Duration.ofMillis(500), Duration.ofSeconds(5));
```

Comment:
It seems this is a best-effort waiting call and we don't have happens-before relation to ensure that the previous task must be done or cancelled.

Filter only `hasLeadership` will proceed `interruptLeadership` seems enough for fixing 518?

Reply:
I thought about the guarantee `LeaderSelector` try to provide, and conclude to that:
* It is crucial for `LeaderSelectorListener.takeLeadership` to not swallow interruption silently without return. Otherwise, states(e.g. `hasLeadership`, `mutex`) could be changed by dated task.
* Interruption for `LeaderSelectorListener.takeLeadership` is a must regardless of leadership. Otherwise, `LeaderSelectorListener.takeLeadership` could run with dated leadership and no future interruption.
* Without `autoRequeue`, `requeue` is hard to use.
   * An attempt could be cancelled due to session state change.
   * `requeue` after `LeaderSelector.interruptLeadership` might be a noop.

For your questions:
1. Yes. It may be better to enforce only one `LeaderSelectorListener.takeLeadership` call simultaneously. But that could make this pr not concentrated.
2. Yes, but `LeaderSelector.interruptLeadership` is not enough as it could introduce dated leadership.

I think I messed up things here, it might be better to make this pr more concentrated on CURATOR-518 and leaves other concerns(concurrency of `LeaderSelector.doWork` and `requeue` usage)  in followups.

Reply:
I agree that we can focus on the fix here. Hard-code timeout seems brittle also. We can give another round for improvement.

Reply:
It is caused by my fix in `cancelElection`. Old code won't cause simultaneous elections, so didn't need such wait.